spacy_experimental/biaffine_parser/tests/test_with_pad_seq_unpad_matrix.py

```diff
@@ -0,0 +1,60 @@
+from typing import Callable, Tuple
+from thinc.api import Model
+
+from spacy_experimental.biaffine_parser.with_pad_seq_unpad_matrix import (
+    InnerInT,
+    InnerOutT,
+    with_pad_seq_unpad_matrix,
+)
+
+
+def _mock_model():
+    return Model("mock_model", _mock_model_forward)
+
+
+def _mock_model_forward(
+    model: Model[InnerInT, InnerOutT], X: InnerInT, is_train: bool
+) -> Tuple[InnerOutT, Callable[[InnerOutT], InnerInT]]:
+    Xf, lens = X
+
+    # Do something resembling our pairwise bilinear layer.
+    Y = Xf @ Xf.transpose(0, 2, 1)  # type: ignore
+
+    def backprop(dY: InnerOutT) -> InnerInT:
+        model.attrs["last_dY"] = dY
+        return Xf + 1.0, lens
+
+    return Y, backprop
+
+
+def test_with_pad_seq_unpad_matrix():
+    model = with_pad_seq_unpad_matrix(_mock_model())
+    X = [
+        model.ops.xp.arange(6, dtype="float32").reshape(3, 2),
+        model.ops.xp.arange(4, dtype="float32").reshape(2, 2),
+    ]
+    Y, backprop = model(
+        X,
+        True,
+    )
+
+    model.ops.xp.testing.assert_equal(
```

Comment:
That's annoying, since these are ragged and need to be list. CuPy has [assert_array_list_equal](https://docs.cupy.dev/en/stable/reference/generated/cupy.testing.assert_array_list_equal.html) for this purpose, which NumPy doesn't have.

So for now I just made it explicit that we are calling this from NumPy.